README.md
# ChainLadder 

ChainLadder is an R package providing methods and models which are typically 
used in insurance claims reserving, including:

- Mack-, Munich- and Bootstrap-chain-ladder models
- General multivariate chain ladder models 
- Loss development factor fitting and Cape Cod models 
- Generalized linear models 
- Utility functions to:
  - convert quickly tables into triangles
  - triangles into tables 
  - cumulative into incremental 
  - incremental into cumulative triangles

## Installation

You can install the stable version from
[CRAN](http://cran.r-project.org/package=ChainLadder):

```s
install.packages('ChainLadder')
```

To install the current development version from github you need the [devtools package](http://cran.r-project.org/web/packages/devtools/index.html) and the other packages on which ChainLadder depends:

```s
install.packages(c("systemfit", "actuar", "Hmisc", "statmod", "tweedie", "cplm"))
```

To install ChainLadder run:
```s
library(devtools)
install_github("mages/ChainLadder")
```

## Usage

```s
library(ChainLadder)
?ChainLadder
demo(ChainLadder)
```

See the ChainLadder package [vignettes](http://cran.r-project.org/web/packages/ChainLadder/) for more details. 

## Citation

To cite package 'ChainLadder' in publications use:

  Markus Gesmann, Daniel Murphy and Wayne Zhang (2014). ChainLadder:
  Statistical methods and models for the calculation of outstanding claims
  reserves in general insurance. R package version 0.1.9.
  
See also:

  Markus Gesmann. Claims Reserving and IBNR. Computational Actuarial Science
  with R. 2014. Chapman and Hall/CRC

## License

This package is free and open source software, licensed under [GPL](https://www.gnu.org/copyleft/gpl.html).

<a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/deed.en_GB"><img alt="Creative Commons Licence" class="c1" src="http://i.creativecommons.org/l/by-sa/4.0/80x15.png" /></a><br />
<span>ChainLadder documentation</span> by Markus Gesmann, Dan Murphy and Wayne Zhang is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/deed.en_GB">Creative Commons Attribution-ShareAlike 4.0 International License</a>. 

NEWS
Version 0.1.9 [2014-12-20]
--------------------------

Changes

  o Updated README and DESCRIPTION file to comply with changes of 
    CRAN policy.


Version 0.1.8 [2014-08-22]
--------------------------

Bug Fixes
  
  o BootChainLadder produced warnings for triangles that had static 
    developments when the argument process.distr was set to "od.pois".
  o as.triangle.data.frame didn't work for a data.frame less than three rows
  o Arguments xlab and ylab were not passed through in plot.triangle, when
    lattice=TRUE

Version 0.1.7 [2013-09-15]
--------------------------

Changes
  
  o The glmReserve function currently doesn't allow the parameter var.power
    to be set to NULL, which would have called the cpglm function of the 
    cplm package. The cplm package is due to dependency issues with lme4 
    no longer available via CRAN.

Version 0.1.6 [2013-08-09]
--------------------------

NEW FEATURES

  o A new function, CLFMdelta, finds the value of 
    delta such that the model coefficients resulting from 
    the 'chainladder' function with that value for argument delta are 
    consistent with an input vector of 'selected' age-to-age factors,
    subject to restrictions on the 'selected' factors relative to
    the input 'Triangle'. See the paper 
    "A Family of Chain-Ladder Factor Models for Selected Link Ratios"
    by Bardis, Majidi, Murphy:
    http://www.variancejournal.org/issues/?fa=article&abstrID=6943
  
  o A new 'coef' method returns the age-to-age factor coefficients of 
    the regression models estimated by the 'chainladder' function.
    
  o Exports a function "LRfunction" that calculates a Triangle's 
    link ratio function and can be used to plot the space of 
    "reasonable link ratio selections" per the CLFM paper.
    
Changes

   o Removed some package dependencies in DESCRIPTION and moved them 
     to Imports.


Version 0.1.5-6 [2013-03-16]
----------------------------

NEW FEATURES

  o The list output of the MackChainLadder function now includes
    the parameter risk and process risk breakdowns of the total risk
    estimate for the sum of projected losses across all origin years
    by development age.
  o The Mack Method's recursive parameter risk calculation now enables 
    Dr. Mack's original two-term formula (the default) and optionally
    the three-term formula found in Murphy's 1994 paper and in the 
    2006 paper by Buchwalder, Buhlmann, Merz, and Wuthrich.
  o A few more Mack Method examples.


Version 0.1.5-5 [2013-02-13]
----------------------------

Bug Fixes

  o The phi-scaling factor in BootChainLadder was incorrect. 
    Instead of calculating the number of data items in the upper left
    triangle as n*(n+1)/2, n*(n-1)/2 was used. Thanks to Thomas
    Girodot for reporting this bug.


Version 0.1.5-4 [2012-11-10]
----------------------------

NEW FEATURES

  o The function "getLatestCumulative" adds attributes to 
    the result
    - names = origin (rownames) from the Triangle
    - rowsname = name of row dimension of Triangle
    - colnames = dev (colnames) from Triangle
    - colsname = name of the column dimension of Triangle
    The function has an additional argument, na.values, a 
    vector of values (e.g., zero) that are synonymous with NA 
    when searching for the rightmost non-NA value
  o as.triangle.data.frame now aggregates multiple data.frame
    records when more than one (origin, dev) observation is

R/Triangles.R
## Author: Markus Gesmann
## Copyright: Markus Gesmann, markus.gesmann@gmail.com
## Date:10/11/2007; 17/09/2008; 16/11/2009

incr2cum <- function(Triangle, na.rm=FALSE){
    if(na.rm){
        upper <- col(Triangle) <= ncol(Triangle) + 1 - row(Triangle)
        upperna <- which(is.na(Triangle[upper]), arr.ind=TRUE)

        Triangle[upper][upperna] <- 0
    }
    cum <- t(apply(Triangle,1, cumsum))
    dimnames(cum) <- dimnames(Triangle)
    class(cum) <- c("triangle", "matrix")
    cum
}


cum2incr <- function(Triangle){
    incr <- cbind(Triangle[,1], t(apply(Triangle,1,diff)))
    dimnames(incr) <- dimnames(Triangle)
    class(incr) <- c("triangle", "matrix")
    incr
}


as.triangle <- function(Triangle, origin="origin", dev="dev", value="value",...){
  UseMethod("as.triangle")
}

as.triangle.matrix <- function(Triangle, origin="origin", dev="dev", value="value",...){
 class(Triangle) <- c("triangle", "matrix")
 if(is.null(dimnames(Triangle))){
     dimnames(Triangle) <- list(origin=1:nrow(Triangle), dev=1:ncol(Triangle))
 }
 names(dimnames(Triangle)) <- c(origin, dev)

 if(is.null(dimnames(Triangle)$origin)){
     dimnames(Triangle)$origin <- 1:nrow(Triangle)
 }
 if(is.null(dimnames(Triangle)$dev)){
     dimnames(Triangle)$dev <- 1:col(Triangle)
 }


 return(Triangle)
}

as.triangle.data.frame <- function(Triangle, origin="origin", dev="dev", value="value", ...){
    d <- dim(Triangle)
    if(length(d) == 2 & d[1]==d[2]){
        matrixTriangle <- as.matrix(Triangle)
        matrixTriangle <- as.triangle(matrixTriangle)
    }else{
        matrixTriangle <- .as.MatrixTriangle(Triangle, origin, dev, value)
    }
    class(matrixTriangle) <- c("triangle", "matrix")
    return(matrixTriangle)
}

as.data.frame.triangle <- function(x, row.names=NULL, optional, lob=NULL, na.rm=FALSE,...){
    longTriangle <- .as.LongTriangle(x, na.rm)
    if(is.null(row.names))
        rownames(longTriangle) <- paste(longTriangle$origin, longTriangle$dev, sep="-")
    if(!is.null(lob))
        longTriangle$lob=lob

    class(longTriangle) <- c("long.triangle", "data.frame")
    return(longTriangle)
}

plot.triangle <- function(x,type="b",xlab="dev. period",ylab=NULL, lattice=FALSE,...){
    .x <- x
    class(.x) <- "matrix"
    if(!lattice){
        matplot(t(.x),type=type,
                xlab=xlab,
                ylab=ifelse(is.null(ylab), deparse(substitute(x)), ylab),...)
    }else{
        df <- as.data.frame(as.triangle(.x))
        xyplot(value ~ dev | factor(origin), data=df, type="l", as.table=TRUE,...)
    }
}

print.triangle <- function(x, ...) {
    class(x) <- tail(class(x), -1)
    NextMethod(x, ...)
    }

.as.MatrixTriangle <- function(x, origin="origin", dev="dev", value="value"){
    ## x has to be a data.frame with columns: origin, dev and value
    x <- x[,c(origin, dev, value)]
    x <- x[order(x[origin], x[dev]),]
    names(x) <- c("origin", "dev", "value")
    .names <- apply(x[,c("origin", "dev", "value")], 2, unique)
    if(class(.names) != "list"){
        .names <- as.list(as.data.frame(.names))
    }
    .namesOD <- .names[c("origin", "dev")]
    ## Expand to include entire array, in case don't have complete data
    .id <- paste(x$origin, x$dev,  sep='.')
    .grid <- expand.grid(.namesOD)
    .grid$id <- paste(.grid$origin, .grid$dev, sep='.')
    .grid$data <- x$value[match(.grid$id, .id)]
    ## Create data array
    .data <- array(.grid$data, dim=unlist(lapply(.namesOD, length)),
                   dimnames=.namesOD)
    return(.data)
}

.as.LongTriangle <- function(Triangle, na.rm=FALSE){
    x <- Triangle
    lx <- expand.grid(origin=as.numeric(dimnames(x)$origin), dev=as.numeric(dimnames(x)$dev))
    ##    lx <- expand.grid(origin=dimnames(x)$origin, dev=dimnames(x)$dev)
    lx$value <- as.vector(x)
    if(na.rm){
        lx <- na.omit(lx)
    }
    return(lx)
}

# Idea: think about a class triangles, which stores an array of triangles, e.g. for several lines of business


R/MackChainLadderFunctions.R

## Author: Markus Gesmann
## Copyright: Markus Gesmann, markus.gesmann@gmail.com
## Date:10/11/2007
## Date:08/09/2008
## Date:22/03/2009
## Date:25/05/2010 Dan

MackChainLadder <- function(Triangle,
                            weights=1,
                            alpha=1,
                            est.sigma="log-linear",
                            tail=FALSE,
                            tail.se=NULL,
                            tail.sigma=NULL)
{
    ## idea: have a list for tail factor
    ## tail=list(f=FALSE, f.se=NULL, sigma=NULL, F.se=NULL)
    ##

    Triangle <- checkTriangle(Triangle)
    m <- dim(Triangle)[1]
    n <- dim(Triangle)[2]

    ## Create chain ladder models

    ## Mack uses alpha between 0 and 2 to distinguish
    ## alpha = 0 ordinary regression with intercept 0
    ## alpha = 1 historical chain ladder age-to-age factors
    ## alpha = 2 straight averages

    ## However, in Zehnwirth & Barnett they use the notation of delta, whereby delta = 2 - alpha
    ## the delta is than used in a linear modelling context.
    delta <- 2-alpha
    CL <- chainladder(Triangle, weights=weights, delta=delta)
    alpha <- 2 - CL$delta

    ## Predict the chain ladder model
    FullTriangle <- predict.TriangleModel(list(Models=CL[["Models"]], Triangle=Triangle))
    ## Estimate the standard error for f and F
    StdErr <- Mack.S.E(CL[["Models"]], FullTriangle, est.sigma=est.sigma,
                       weights=CL[["weights"]], alpha=alpha)


    Total.SE <- TotalMack.S.E(FullTriangle, StdErr$f, StdErr$f.se, StdErr$F.se)

    ## Check for tail factor
    if(is.logical(tail)){
        if(tail){
            tail <- tailfactor(StdErr$f)
            tail.factor <- tail$tail.factor
        }else{
            tail.factor <- tail
        }
    }else{
        if(is.numeric(tail))
            tail.factor <- tail
    }


    ## Estimate standard error for the tail factor f and F
    if(tail.factor>1){
        tail.out <- tail.SE(FullTriangle, StdErr, Total.SE, tail.factor,
                            tail.se=tail.se, tail.sigma=tail.sigma)

        FullTriangle <- tail.out[["FullTriangle"]]
        StdErr <- tail.out[["StdErr"]]
        Total.SE <- tail.out[["Total.SE"]]
    }else{
        Total.SE <- TotalMack.S.E(FullTriangle, StdErr$f, StdErr$f.se, StdErr$F.se)
    }

    ## Add process and parameter risk
    StdErr <- c(StdErr, MackRecursive.S.E(FullTriangle, StdErr$f, StdErr$f.se, StdErr$F.se))

    ## Collect the output
    output <- list()
    output[["call"]] <-  match.call(expand.dots = FALSE)
    output[["Triangle"]] <- Triangle
    output[["FullTriangle"]] <- FullTriangle
    output[["Models"]] <- CL[["Models"]]
    output[["f"]] <- StdErr$f
    output[["f.se"]] <- StdErr$f.se
    output[["F.se"]] <- StdErr$F.se
    output[["sigma"]] <- StdErr$sigma
    output[["Mack.ProcessRisk"]]   <- StdErr$FullTriangle.procrisk  # new dmm
    output[["Mack.ParameterRisk"]] <- StdErr$FullTriangle.paramrisk  # new dmm
    output[["Mack.S.E"]] <- sqrt(StdErr$FullTriangle.procrisk^2 +StdErr$FullTriangle.paramrisk^2)
    output[["weights"]] <- CL$weights
    output[["alpha"]] <- alpha
    ## total.procrisk <- apply(StdErr$FullTriangle.procrisk, 2, function(x) sqrt(sum(x^2)))
    output[["Total.Mack.S.E"]] <- Total.SE
    output[["tail"]] <- tail
    class(output) <- c("MackChainLadder", "TriangleModel", "list")
    return(output)
}

##############################################################################
## Calculation of the mean squared error and standard error
## mean squared error = stochastic error (process variance) + estimation error
## standard error = sqrt(mean squared error)

Mack.S.E <- function(MackModel, FullTriangle, est.sigma="log-linear", weights, alpha){
    n <- ncol(FullTriangle)
    m <- nrow(FullTriangle)
    f <- rep(1,n)
    f.se <- rep(0,n)
    sigma <- rep(0,(n-1))

    ## Extract estimated slopes, std. error and sigmas
    f[1:(n-1)] <- sapply(MackModel, function(x) summary(x)$coef["x","Estimate"])
    f.se[1:(n-1)] <- sapply(MackModel, function(x) summary(x)$coef["x","Std. Error"])
    sigma[1:(n-1)] <- sapply(MackModel, function(x) summary(x)$sigma)

    isna <- is.na(sigma)
    ## Think about weights!!!

    if(est.sigma %in% "log-linear"){
        ## estimate sigma[n-1] via log-linear regression
        sig.model <- estimate.sigma(sigma)
        sigma <- sig.model$sigma

        p.value.of.model <- summary(sig.model$model)$coefficient[2,4]
        if(p.value.of.model > 0.05){
            warning(paste("'loglinear' model to estimate sigma_n doesn't appear appropriate.",
                          "\np-value > 5.\n",
                          "est.sigma will be overwritten to 'Mack'.\n",
                          "Mack's estimation method will be used instead."))

            est.sigma <- "Mack"
        }else{
            f.se[isna] <- sigma[isna]/sqrt(weights[1,isna]*FullTriangle[1,isna]^alpha[isna])
        }
    }
    if(est.sigma %in% "Mack"){
        for(i in which(isna)){   # usually i = n - 1
            sigma[i] <- sqrt(abs(min((sigma[i - 1]^4/sigma[i - 2]^2),
                                     min(sigma[i - 2]^2, sigma[i - 1]^2))))
            f.se[i] <- sigma[i]/sqrt(weights[1,i]*FullTriangle[1,i]^alpha[i])
        }
    }
    if(is.numeric(est.sigma)){
        for(i in seq(along=est.sigma)){
            l <- length(est.sigma)
            sigma[n-i] <- est.sigma[l-i+1]
            f.se[n-i] <- sigma[n-i]/sqrt(weights[1,n-i]*FullTriangle[1,n-i]^alpha[n-i])
        }
    }

    W <- weights
    W[is.na(W)] <- 1
    F.se <- t(sigma/t(sqrt(W[,-n]*t(t(FullTriangle[,-n])^alpha[-n]))))

    return(list(sigma=sigma,
                f=f,
                f.se=f.se,
                F.se=F.se)
           )
}
################################################################
MackRecursive.S.E <- function(FullTriangle, f, f.se, F.se){

    nn <- length(f)
    n <- ncol(FullTriangle)
    m <- nrow(FullTriangle)

    FullTriangle.procrisk <- FullTriangle[,1:nn] * 0
    FullTriangle.paramrisk <- FullTriangle[,1:nn] * 0

    ## Recursive Formula
    rowindex <- 2:m
    if(m>nn)
        rowindex <- c((m-nn+1):m)
    for(i in rowindex){
        for(k in c((nn+1-i):(nn-1))){
            if(k>0) {
		FullTriangle.procrisk[i,k+1] <- sqrt(
                               FullTriangle[i,k]^2*(F.se[i,k]^2)
                               + FullTriangle.procrisk[i,k]^2*f[k]^2
			       )
		FullTriangle.paramrisk[i,k+1] <- sqrt(
                               FullTriangle[i,k]^2*(f.se[k]^2)
                               + FullTriangle.paramrisk[i,k]^2*f[k]^2
			       )
            }
    	}
    }

     if(f[nn] > 1){ ## tail factor > 1
        k <- nn
        Tail.procrisk <- sqrt(
                              FullTriangle[,k]^2*(F.se[,k]^2)
                              + FullTriangle.procrisk[,k]^2*f[k]^2)
        FullTriangle.procrisk <- cbind(FullTriangle.procrisk, Tail.procrisk)

        Tail.paramrisk <- sqrt(
                               FullTriangle[,k]^2*(f.se[k]^2)
                               + FullTriangle.paramrisk[,k]^2*f[k]^2)
        FullTriangle.paramrisk <- cbind(FullTriangle.paramrisk,Tail.paramrisk)
    }



    return(list(FullTriangle.procrisk=FullTriangle.procrisk,
                FullTriangle.paramrisk=FullTriangle.paramrisk))
}

################################################################################
## Total reserve SE

TotalMack.S.E <- function(FullTriangle,f, f.se, F.se){

    C <- FullTriangle
    n <- ncol(C)
    m <- nrow(C)

    total.seR <- 0*c(1:n)

    for(k in c(1:(n-1))){
        total.seR[k+1] <- sqrt(total.seR[k]^2 * f[k]^2 +
                               sum(C[c((m+1-k):m),k]^2 *
                                   (F.se[c((m+1-k):n),k]^2),na.rm=TRUE)
                               + sum(C[c((m+1-k):m),k],na.rm=TRUE)^2 * f.se[k]^2 )
    }
    return(total.seR[length(total.seR)])
}

##############################################################################

estimate.sigma <- function(sigma){
    if(!all(is.na(sigma))){
        n <- length(sigma)
        dev <- 1:n
        my.dev <- dev[!is.na(sigma) & sigma > 0]
        my.model <- lm(log(sigma[my.dev]) ~ my.dev)
        sigma[is.na(sigma)] <- exp(predict(my.model, newdata=data.frame(my.dev=dev[is.na(sigma)])))
    }

    return(list(sigma=sigma, model=my.model))
}


########################################################################
## Estimate standard error for tail

tail.SE <- function(FullTriangle, StdErr, Total.SE, tail.factor, tail.se=NULL, tail.sigma=NULL){
    n <- ncol(FullTriangle)
    m <- nrow(FullTriangle)

    FullTriangle <- cbind(FullTriangle, FullTriangle[,n] * tail.factor)
    dimnames(FullTriangle) <- list(origin=dimnames(FullTriangle)[[1]],
                                   dev=c(dimnames(FullTriangle)[[2]][1:n], "Inf"))

    StdErr$f[n] <- tail.factor

    ## Idea: linear model for f, estimate dev for tail factor
    ## linear model for f.se and sigma and put dev from above in
    start <- 1
    .f <- StdErr$f[start:(n-1)]
    .dev <- c(start:(n-1))
##    mf <- lm(log(.f[.f>1]-1) ~ .dev[.f>1])
    mf <- lm(log(.f-1) ~ .dev)
    tail.pos <- ( log(StdErr$f[n]-1) - coef(mf)[1] ) / coef(mf)[2]

    if(is.null(tail.se)){
        .fse <- StdErr$f.se[start:(n-1)]
        mse <- lm(log(.fse) ~ .dev)
        tail.se <- exp(predict(mse, newdata=data.frame(.dev=tail.pos)))
    }
    StdErr$f.se[n] <- tail.se

    if(is.null(tail.sigma)){
        .sigma <- StdErr$sigma[start:(n-1)]
        msig <- lm(log(.sigma) ~ .dev)
        tail.sigma <- exp(predict(msig, newdata=data.frame(.dev=tail.pos)))
    }
    StdErr$sigma <- c(StdErr$sigma, tail.sigma=as.numeric(tail.sigma))

    ## estimate the stanard error of the tail factor ratios
    se.F.tail <- tail.sigma/sqrt(FullTriangle[,n])
    StdErr$F.se <- cbind(StdErr$F.se, se.F.tail)

    Total.SE <- sqrt(Total.SE^2 * StdErr$f[n]^2
                     + sum(FullTriangle[c(1:m), n]^2 * (StdErr$F.se[c(1:m), n]^2), na.rm=TRUE)
                     + sum(FullTriangle[c(1:m), n], na.rm=TRUE)^2 * StdErr$f.se[n]^2)

    output <- list(FullTriangle=FullTriangle, StdErr=StdErr, Total.SE=Total.SE)
    return(output)
}




##############################################################################
## Summary
##
summary.MackChainLadder <- function(object,...){
    ## Summarise my results
    Latest <- getLatestCumulative(object$Triangle)

    ex.origin.period <- Latest!=0

    Ultimate <- object[["FullTriangle"]][,ncol(object[["FullTriangle"]])]
    Dev.To.Date <- Latest/Ultimate
    IBNR <- Ultimate-Latest
    Mack.S.E <- object[["Mack.S.E"]][,ncol(object[["Mack.S.E"]])]
    CV <- Mack.S.E/(Ultimate-Latest)

    ByOrigin <- data.frame(Latest, Dev.To.Date, Ultimate, IBNR, Mack.S.E, CV)
    names(ByOrigin)[6]="CV(IBNR)"
    ByOrigin <- ByOrigin[ex.origin.period,]

    Totals <-  c(sum(Latest,na.rm=TRUE),
                 sum(Latest,na.rm=TRUE)/sum(Ultimate,na.rm=TRUE),
                 sum(Ultimate,na.rm=TRUE),
                 sum(IBNR,na.rm=TRUE), object[["Total.Mack.S.E"]],
                 object[["Total.Mack.S.E"]]/sum(IBNR,na.rm=TRUE)
                 )
    # Totals <- c(Totals, round(x[["Total.Mack.S.E"]]/sum(res$IBNR,na.rm=TRUE),2))
    Totals <- as.data.frame(Totals)

    colnames(Totals)=c("Totals")
    rownames(Totals) <- c("Latest:","Dev:","Ultimate:",
                          "IBNR:","Mack S.E.:",
                          "CV(IBNR):")

    output <- list(ByOrigin=ByOrigin, Totals=Totals)
    return(output)
}

getLatestCumulative <- function(cumulative.tri)
  apply(cumulative.tri, 1L, function(x) ifelse(length(w <- which(!is.na(x))) > 0L, x[tail(w, 1L)], x[1L]))

##############################################################################
## print
##
print.MackChainLadder <- function(x,...){

    summary.x <- summary(x)
    print(x$call)
    cat("\n")
    print(format(summary.x$ByOrigin, big.mark = ",", digits = 3),...)

    Totals <- summary.x$Totals
    Totals[1:5,] <- formatC(Totals[1:5,], big.mark=",",digits=2,format="f")
    cat("\n")
    print(Totals, quote=FALSE)
    #invisible(x)

}


################################################################################
## plot
##
plot.MackChainLadder <- function(x, mfrow=c(3,2), title=NULL,lattice=FALSE,...){


    .myResult <-  summary(x)$ByOrigin

    .FullTriangle <- x[["FullTriangle"]]
    .Triangle <- x[["Triangle"]]

    if(!lattice){
        if(is.null(title)) myoma <- c(0,0,0,0) else myoma <- c(0,0,2,0)

        op=par(mfrow=mfrow, oma=myoma, mar=c(4.5,4.5,2,2))

        plotdata <- t(as.matrix(.myResult[,c("Latest","IBNR")]))
        n <- ncol(plotdata)

if(getRversion() < "2.9.0") { ## work around missing feature

        bp <- barplot(plotdata,
                      legend.text=c("Latest","Forecast"),
                      ##    args.legend=list(x="topleft"), only avilable from R version >= 2.9.0
                      names.arg=rownames(.myResult),
                      main="Mack Chain Ladder Results",
                      xlab="Origin period",
                      ylab="Value",#paste(Currency,myUnit),
                      ylim=c(0, max(apply(.myResult[c("Ultimate", "Mack.S.E")],1,sum),na.rm=TRUE)))

    }else{
   bp <- barplot(plotdata,
                      legend.text=c("Latest","Forecast"),
                      args.legend=list(x="topleft"),
                      names.arg=rownames(.myResult),
                      main="Mack Chain Ladder Results",
                      xlab="Origin period",
                      ylab="Value",#paste(Currency,myUnit),
                      ylim=c(0, max(apply(.myResult[c("Ultimate", "Mack.S.E")],1,sum),na.rm=TRUE)))
    }
        ## add error ticks
        require("Hmisc")
        errbar(x=bp, y=.myResult$Ultimate,
               yplus=(.myResult$Ultimate + .myResult$Mack.S.E),
               yminus=(.myResult$Ultimate - .myResult$Mack.S.E),
               cap=0.05,
               add=TRUE)

        matplot(t(.FullTriangle), t="l",
                main="Chain ladder developments by origin period",
                xlab="Development period",
                ylab="Amount", #paste(Currency, myUnit)
                )
        matplot(t(.Triangle), add=TRUE)

        Residuals=residuals(x)
        plot(standard.residuals ~ fitted.value, data=Residuals,
             ylab="Standardised residuals", xlab="Fitted")
        lines(lowess(Residuals$fitted.value, Residuals$standard.residuals), col="red")
        abline(h=0, col="grey")
        plot(standard.residuals ~ origin.period, data=Residuals,
             ylab="Standardised residuals", xlab="Origin period")
        lines(lowess(Residuals$origin.period, Residuals$standard.residuals), col="red")
        abline(h=0, col="grey")
        plot(standard.residuals ~ cal.period, data=Residuals,
             ylab="Standardised residuals", xlab="Calendar period")
        lines(lowess(Residuals$cal.period, Residuals$standard.residuals), col="red")
        abline(h=0, col="grey")
        plot(standard.residuals ~ dev.period, data=Residuals,
             ylab="Standardised residuals", xlab="Development period")
        lines(lowess(Residuals$dev.period, Residuals$standard.residuals), col="red")
        abline(h=0, col="grey")

        title( title , outer=TRUE)
        par(op)

    }else{
        long <- expand.grid(origin=as.numeric(dimnames(.FullTriangle)$origin),
                            dev=as.numeric(dimnames(.FullTriangle)$dev))
        long$value <- as.vector(.FullTriangle)
        long$valuePlusMack.S.E <-  long$value + as.vector(x$Mack.S.E)
        long$valueMinusMack.S.E <- long$value - as.vector(x$Mack.S.E)
        xyplot(valuePlusMack.S.E + valueMinusMack.S.E + value ~ dev |
               factor(origin), data=long[!is.na(long$value),], type="l", lty=c(3,3,1), as.table=TRUE,
               main="Chain ladder developments by origin period",
               xlab="Development period",
               ylab="Amount",col=1,
               key=list(lines=list(lty=c(1,3), col=1),
               text=list(lab=c("Chain ladder dev.", "Mack's S.E.")),
               space="top", columns=2),...)
    }
}
################################################################################
## residuals
##
residuals.MackChainLadder <- function(object,...){
    m <- nrow(object[["Triangle"]])
    n <- ncol(object[["Triangle"]])
    myresiduals <- unlist(lapply(object[["Models"]], resid,...))
   
    standard.residuals <- rep(NA, length(myresiduals))
    
    ## Identify if we have standardized residuals for all items, 
    ## e.g. this is not the case if some weights have been set to zero
    x=lapply(lapply(object$Model, resid), function(x) as.numeric(names(x)))
	y=lapply(lapply(object$Model, rstandard), function(x) as.numeric(names(x)))
	names(y)=1:length(y)
	names(x)=1:length(x)		
    rst <- unlist(lapply(names(x), function(z) x[[z]] %in% y[[z]]))
    ## rst holds the indices with standardised residuals for weights > 0 
    
    standard.residuals[rst] <- unlist(lapply(object[["Models"]], rstandard,...))
    fitted.value <- unlist(lapply(object[["Models"]], fitted))
    origin.period <- as.numeric(unlist(lapply(lapply(object[["Models"]],residuals),names)))
    dev.period <-rep(1:(n-1), sapply(lapply(object[["Models"]],residuals),length))
    cal.period <- origin.period + dev.period - 1

    myResiduals <- data.frame(origin.period,
                              dev.period,
                              cal.period,
                              residuals=myresiduals,
                              standard.residuals,
                              fitted.value)

    return(na.omit(myResiduals))
}


Mack <- function(triangle, weights=1, alpha=1){
    ## Mack Chain Ladder:
    ## triangle: cumulative claims triangle
    ## weights : weights
    ## alpha=0 : gives straight average of the chain ladder age-to-age factors
    ## alpha=1 : gives the historical chain ladder age-to-age factors
    ## alpha=2 : is the result of an ordinary regression of C_{i,k+1} against C_{i,k} with intercept 0.
    weights <- checkWeights(weights, triangle)

    m <- nrow(triangle)
    n <- ncol(triangle)

    ## Individual chain ladder age-to-age factors:
    F <- triangle[,-1]/triangle[,-n]
#    if(is.numeric(weights)){
#        weights2 <- triangle
#        weights2[!is.na(weights2)] <- weights
#        weights <- weights2
#    }

    wCa <- weights[,-n] * triangle[,-n]^alpha
    ## Note NA^0=1, hence set all cells which are originally NA back to NA
    wCa[is.na(triangle[,-n])] <- NA

    wCaF <- wCa*F
    ## Set diagonal of wCa to NA
    wCa[row(wCa)==n+1-col(wCa)] <- NA
    ## Chain-ladder age-to-age factors
    f <- apply( wCaF, 2, sum, na.rm=TRUE)/apply( wCa, 2, sum, na.rm=TRUE)

    ## Get full triangle
    avDFs <- c(f,1)
    dim(avDFs) <- c(1,n,1)
    ultDFs <- getUltDFs(avDFs)

    Latest <- getLatestCumulative(triangle)
    ults <- getUltimates(Latest, ultDFs)
    FullTriangle <- getExpected(ults, 1/ultDFs)
    dim(FullTriangle)=c(n,n)

    ## Estimate standard errors

    k <- c(1:(n-2))
    wCa_F_minus_f <- weights[,k] * triangle[,k]^alpha * t(t(F)-f)[,k]^2

    wCa_F_minus_f <-  triangle[,k]^alpha * t(t(F)-f)[,k]^2

    sigma <- sqrt( 1/c(n-k-1) * apply(wCa_F_minus_f, 2, sum, na.rm=TRUE) )
    # Mack approximation for sigma_{n-1}
    sigma_n1 <- sqrt( abs(min(sigma[n-2]^4/sigma[n-3]^2, min(sigma[n-3]^2, sigma[n-2]^2)) ))
    sigma <- c(sigma, sigma_n1)

    # Estimate f.se
    wCa2 <-  weights[,c(1:(n-1))] * triangle[,c(1:(n-1))]^alpha
    ## Note NA^0=1, hence set all cells which are originally NA back to NA
    wCa2[is.na(triangle[,-n])] <- NA

    ## Set diagonal of wCa to na
    wCa2[row(wCa2)==n+1-col(wCa2)] <- NA
    f.se <- sigma / sqrt( apply(wCa2, 2, sum, na.rm=TRUE))

    # Estimate F.se
    W <- weights
    W[is.na(W)] <- 1
    F.se <- t(sigma/sqrt(t(W[,c(1:(n-1))]*FullTriangle[,c(1:(n-1))]^alpha)))
#    F.se <- t(sigma/sqrt(t(FullTriangle[,c(1:(n-1))]^alpha)))

    F.se[is.na(FullTriangle[,c(1:(n-1))])] <- NA

    FullTriangle.se <- FullTriangle * 0
    ## Recursive Formula
    rowindex <- 2:m
    if(m>n)
        rowindex <- c((m-n+1):m)
    for(i in rowindex){
        for(k in c((n+1-i):(n-1))){
            if(k>0)
                FullTriangle.se[i,k+1] = sqrt(
                               FullTriangle[i,k]^2*(F.se[i,k]^2+f.se[k]^2) #
                               + FullTriangle.se[i,k]^2*f[k]^2
                               )
    	}
    }

    output <- list()
    output[["call"]] <-  match.call(expand.dots = FALSE)
    output[["Triangle"]] <- triangle
    output[["Latest"]] <- Latest
    output[["FullTriangle"]] <- FullTriangle
    output[["f"]] <- f
    output[["F"]] <- F
    output[["f.se"]] <- f.se
    output[["F.se"]] <- F.se
    output[["sigma"]] <-sigma
    output[["Mack.S.E"]] <- FullTriangle.se
    class(output) <- c("Mack","list")
    return(output)
}

print.Mack <- function(x,...){
    n <- ncol(x$Triangle)
    df <- data.frame(Latest=x$Latest, f=c(NA, x$f), f.se=c(NA, x$f.se),
                     Ultimate=x$FullTriangle[,n])
    df$IBNR <- df$Ultimate-df$Latest
    df$Mack.S.E=x$Mack.S.E[,n]
    df$CV=df$Mack.S.E/df$IBNR
    print(df)
}
